### PR TITLE
Show scheduling errors with their respective fields

### PR DESCRIPTION
**Pull request by kevindew**
Trello: https://trello.com/c/rGjDviSL/848-adapt-scheduling-implementation-for-iterated-workflow

This allows each of the requirements issues to link directly to a field. It does this by breaking up the datetime errors into ones that are date or time. To complete this the autocomplete component needed updating to accept errors.

There are few things that aren't quite ideal in this which I'll explain:

- The fields for scheduling are most prefixed with `schedule_` this is done to disambiguate them amongst the many fields in the requirements yaml file, it would have been nicer to have date, time and action as names but they looked rather generic in amongst that file;
- The time parsing is quite strict to limit checking that a time input is actually a time, the alternative is pretty wild west parsing wise without further complication - unless anyone has better ideas?
- the date_input doesn't accept error_items as input, hopefully we can fix that soon
- the date_input doesn't have a label inside it so that's not caught in an error state

Screenshot:

Screen Shot 2019-06-05 at 09 50 04

### CODE REVIEW

File: app/services/requirements/scheduled_datetime_checker.rb [hunk 1]
```diff
@@ -2,76 +2,82 @@
...
+          time_period: time_period_for_issue(MINIMUM_FUTURE_TIME_PERIOD),
+        )
+      end
     end
 
-    def too_far_in_the_future?
-      now = Time.zone.now
-      parsed_datetime > now.advance(MAXIMUM_FUTURE_TIME_PERIOD).end_of_day
-    end
+    def check_datetime_is_foreseeable
```
Comment (emmabeynon) [inline app/services/requirements/scheduled_datetime_checker.rb]: I don't think foreseeable is the right word here; it's more about predictability and we've got a concrete future time period that we don't need to predict.

File: app/views/schedule/scheduled_publishing_datetime.html.erb [hunk 2]
```diff
@@ -1,49 +1,53 @@
...
         hint: t("schedule.scheduled_publishing_datetime.date_hint_text"),
-        id: date_input_id,
+        id: "scheduled-date",
+        error_message: error_message.presence,
         items: [
           {
             name: "day",
             width: 2,
-            value: schedule_param["day"] || publish_date&.day
+            value: params.dig(:schedule, :scheduled_date, :day) || publish_time&.day
```
Comment (emmabeynon) [inline app/views/schedule/scheduled_publishing_datetime.html.erb]: I wonder if `publish_time` should be renamed `publish_datetime`?  When I first saw the variable I assumed it was the time only, and got confused when I saw it appear as a date value.

File: app/services/requirements/scheduled_datetime_checker.rb (hunk 1 above)
Comment (kevindew) [reply app/services/requirements/scheduled_datetime_checker.rb]: Sure, any suggestions on the wording? I was hoping I could get something with an is rather than isn't and found isnt_too_far_in_the_future a bit wordy.

somethings considered and rejected: is_realistic is_soonish is_sane

File: app/views/schedule/scheduled_publishing_datetime.html.erb (hunk 2 above)
Comment (kevindew) [reply app/views/schedule/scheduled_publishing_datetime.html.erb]: Sure, I'll have a look at that - it's a bit of a pain as a Time object in Ruby always represents a date so that kind of makes sense to name it a time. But I know we also deal with time as a string.

**Comment by benthorner** [comment]
Was just trying to have a look at this - does it need rebasing?

**Comment by kevindew** [comment]
> Was just trying to have a look at this - does it need rebasing?

Just done so

File: app/views/schedule/scheduled_publishing_datetime.html.erb [hunk 3]
```diff
@@ -1,49 +1,53 @@
...
-<% publish_date = current_scheduled_publish_datetime || default_date %>
-<% schedule_param = params["schedule"] ? params["schedule"] : {} %>
-<% selected_time = schedule_param["time"] ||  current_scheduled_publish_datetime&.strftime("%l:%M%P")&.strip %>
-<% action_not_selected = @issues&.select { |i| i.issue_key == :action_not_selected }&.first&.to_item %>
-
+<%
+  current_scheduled_publish_datetime = @edition.scheduled_publishing_datetime
+  default_time = Time.current.tomorrow
+  publish_time = current_scheduled_publish_datetime || default_time
+  action_not_selected = @issues&.select { |i| i.issue_key == :action_not_selected }&.first&.to_item
```
Comment (benthorner) [inline app/views/schedule/scheduled_publishing_datetime.html.erb]: Looks like it's not part of this PR, but the requirements system has a `items_for` method especially for this :-).

File: app/views/schedule/scheduled_publishing_datetime.html.erb [hunk 4]
```diff
@@ -1,49 +1,53 @@
...
-<% default_date = Time.zone.now.tomorrow %>
-<% publish_date = current_scheduled_publish_datetime || default_date %>
-<% schedule_param = params["schedule"] ? params["schedule"] : {} %>
-<% selected_time = schedule_param["time"] ||  current_scheduled_publish_datetime&.strftime("%l:%M%P")&.strip %>
-<% action_not_selected = @issues&.select { |i| i.issue_key == :action_not_selected }&.first&.to_item %>
-
+<%
+  current_scheduled_publish_datetime = @edition.scheduled_publishing_datetime
+  default_time = Time.current.tomorrow
+  publish_time = current_scheduled_publish_datetime || default_time
```
Comment (benthorner) [inline app/views/schedule/scheduled_publishing_datetime.html.erb]: If we had two variables here - `publish_date` and `publish_time` - could we avoid all the `||`ing we do later on? This might be a follow-up refactor, though.

File: app/services/requirements/scheduled_datetime_checker.rb (hunk 1 above)
Comment (emmabeynon) [reply app/services/requirements/scheduled_datetime_checker.rb]: `within_time_limit`? I appreciate it still has multiple words but I don't think there is a single word to describe this.

File: app/services/requirements/scheduled_datetime_checker.rb [hunk 5]
```diff
@@ -2,76 +2,82 @@
...
 
     def pre_submit_issues
-      issues = []
+      validity_issues = check_datetime_valid
+      return CheckerIssues.new(validity_issues) if validity_issues.any?
 
-      if day.blank? || month.blank? || year.blank?
-        issues << Issue.new(:scheduled_datetime, :invalid, field: "date")
-      end
+      issues = [check_datetime_is_future,
```
[the remark this answers is not in the document]
Comment (kevindew) [reply app/services/requirements/scheduled_datetime_checker.rb]: Would you mind grabbing me a good example?

[the remark this answers is not in the document]
Comment (benthorner) [reply app/services/requirements/scheduled_datetime_checker.rb]: I think [the attachment checker](https://github.com/alphagov/content-publisher/blob/master/app/services/requirements/file_attachment_checker.rb#L93) is pretty good, as it too has some quite complex checks to deal with. Hope that helps.

File: app/views/schedule/scheduled_publishing_datetime.html.erb (hunk 3 above)
Comment (kevindew) [reply app/views/schedule/scheduled_publishing_datetime.html.erb]: Yup I think I squash this later in this PR

File: app/views/schedule/scheduled_publishing_datetime.html.erb (hunk 4 above)
Comment (kevindew) [reply app/views/schedule/scheduled_publishing_datetime.html.erb]: I think that's mostly unavoidable here as we have both user input and a time object. Would prefer looking at outside of PR.

File: spec/services/requirements/scheduled_datetime_checker_spec.rb [hunk 6]
```diff
@@ -1,101 +1,118 @@
...
-      datetime_params[:time] = "def"
+    it "returns an issue if the datetime is too far into the future" do
+      time_period = { days: 1, months: 14 }
+      future_date = Time.current.advance(time_period)
 
-      issue = I18n.t!("requirements.scheduled_datetime.invalid_input.form_message")
-      issues = Requirements::ScheduledDatetimeChecker.new(datetime_params).pre_submit_issues
+      issues = Requirements::ScheduledDatetimeChecker.new(
+        date: { day: future_date.day,
+                          month: future_date.month,
```
Comment (emmabeynon) [inline spec/services/requirements/scheduled_datetime_checker_spec.rb]: Same indentation issue here and line 107.

File: spec/services/requirements/scheduled_datetime_checker_spec.rb [hunk 7]
```diff
@@ -1,101 +1,118 @@
...
 
-    it "returns an issue if the date or time fields are blank" do
-      datetime_params[:day] = ""
-      datetime_params[:time] = ""
+    it "returns a date issue if the date is in the past" do
+      two_days_ago = Time.current - 2.days
+
+      issues = Requirements::ScheduledDatetimeChecker.new(
+        date: { day: two_days_ago.day,
+                          month: two_days_ago.month,
```
[the remark this answers is not in the document]
Comment (kevindew) [reply spec/services/requirements/scheduled_datetime_checker_spec.rb]: Oh good catch!

**Comment by kevindew** [comment]
Just fixed those @emmabeynon
